### PR TITLE
En 6439 hardfork system tests

### PR DESCRIPTION
**Pull request by iulianpascalau**
- fixed hardfork configs
- hardfork trigger now calls Close methods on registered components so the consensus & sync process will stop
- added hardfork script
- trigger now starts the hardfork process only once
- trigger can not be called with an epoch lower than 1
- added instant hardfork message broadcast in the network
- added "copy-back configs" feature on the testnet scripts helping the debug process inside a local testnet, added possibility to compile the node binary with a generated version number
- added a component to check if a hardfork import should occur, the version is taken into account.

### CODE REVIEW

File: scripts/testnet/config.sh [hunk 1]
```diff
@@ -36,3 +36,7 @@ if [ $USE_TXGEN -eq 1 ]; then
   copyTxGenConfig
   updateTxGenConfig
 fi
+
+if [ $USE_HARDFORK -eq 1 ]; then
+  changeConfigForHardfork
+fi
```
Comment (sasurobert) [inline scripts/testnet/config.sh:42]: empty line

File: scripts/testnet/include/config.sh [hunk 2]
```diff
@@ -188,3 +188,10 @@ updateJSONValue() {
 
   sed -i "s,\"$key\": .*\$,\"$key\": $escaped_value\,," $filename
 }
+
+changeConfigForHardfork(){
+  pushd $TESTNETDIR/node/config
+
+  export FIRST_PUBKEY=$(cat nodesSetup.json | grep pubkey -m 1 | cut -c18-209)
+  updateTOMLValue config.toml "PublicKeyToListenFrom" "\"$FIRST_PUBKEY\""
+}
```
Comment (sasurobert) [inline scripts/testnet/include/config.sh:197]: add empty line

File: scripts/testnet/variables.sh [hunk 3]
```diff
@@ -145,3 +145,4 @@ export TOTAL_OBSERVERCOUNT=$total_observer_count
 let "total_node_count = $SHARD_VALIDATORCOUNT * $SHARDCOUNT + $META_VALIDATORCOUNT + $TOTAL_OBSERVERCOUNT"
 export TOTAL_NODECOUNT=$total_node_count
 
+export USE_HARDFORK=1
```
Comment (sasurobert) [inline scripts/testnet/variables.sh:148]: add empty line

File: update/errors.go [hunk 4]
```diff
@@ -211,3 +211,6 @@ var ErrNilChanStopNodeProcess = errors.New("nil channel to stop node")
 
 // ErrNilEpochConfirmedNotifier signals that nil epoch confirmed notifier was provided
 var ErrNilEpochConfirmedNotifier = errors.New("nil epoch confirmed notifier")
+
+// ErrTriggerAlreadyDone signals that the trigger is already in action, can not re-enter
```
Comment (SebastianMarian) [inline update/errors.go:215]: ErrTriggerAlreadyInAction ?

File: update/interface.go [hunk 5]
```diff
@@ -227,3 +227,9 @@ type EpochChangeConfirmedNotifier interface {
 	RegisterForEpochChangeConfirmed(handler func(epoch uint32))
 	IsInterfaceNil() bool
 }
+
+// Closer defines the functionality of an entity that can be closed
```
Comment (SebastianMarian) [inline update/interface.go:231]: I already defined one in core/close. Maybe delete one of them.

File: scripts/testnet/config.sh (hunk 1 above)
Comment (iulianpascalau) [reply scripts/testnet/config.sh:42]: done

File: scripts/testnet/include/config.sh (hunk 2 above)
Comment (iulianpascalau) [reply scripts/testnet/include/config.sh:197]: done

File: scripts/testnet/variables.sh (hunk 3 above)
Comment (iulianpascalau) [reply scripts/testnet/variables.sh:148]: done

File: update/errors.go (hunk 4 above)
Comment (iulianpascalau) [reply update/errors.go:215]: done

File: update/interface.go (hunk 5 above)
Comment (iulianpascalau) [reply update/interface.go:231]: will link to that one. Having only primitive types in its description makes this interface "redeclarable" in all other packages.

File: update/errors.go [hunk 6]
```diff
@@ -220,3 +220,9 @@ var ErrInvalidTimeToWaitAfterHardfork = errors.New("invalid time to wait after h
 
 // ErrInvalidEpoch signals that an invalid epoch has been provided
 var ErrInvalidEpoch = errors.New("invalid epoch")
+
+// ErrNilImportStartHandler signals that a nil import start handler has been provided
+var ErrNilImportStartHandler = errors.New("nil imports start handler")
```
Comment (SebastianMarian) [inline update/errors.go:225]: import

File: process/sync/storageBootstrap/baseStorageBootstrapper.go [hunk 7]
```diff
@@ -71,7 +71,10 @@ func (st *storageBootstrapper) loadBlocks() error {
 	round := st.bootStorer.GetHighestRound()
 	if round == 0 {
 		log.Debug("Load blocks does nothing as start from genesis")
-		return nil
+		err = st.bootStorer.SaveLastRound(0)
```
Comment (SebastianMarian) [inline process/sync/storageBootstrap/baseStorageBootstrapper.go:74]: I think this is not necessary as anyway 0 is the default value. Also you can remove lines 73 and 75

File: update/errors.go (hunk 6 above)
Comment (sasurobert) [reply update/errors.go:228]: done

File: update/process/metaBlock.go [hunk 8]
```diff
@@ -98,9 +98,10 @@ func (m *metaBlockCreator) CreateNewBlock(
 		SoftwareVersion:        []byte(""),
 		AccumulatedFees:        big.NewInt(0),
 		AccumulatedFeesInEpoch: big.NewInt(0),
+		DeveloperFees:          big.NewInt(0),
```
Comment (SebastianMarian) [inline update/process/metaBlock.go:101]: DevFees or rename the next one DeveloperFeesInEpoch

File: update/process/metaBlock.go [hunk 9]
```diff
@@ -98,9 +98,10 @@ func (m *metaBlockCreator) CreateNewBlock(
...
+		DevFeesInEpoch:         big.NewInt(0),
```
Comment (SebastianMarian) [inline update/process/metaBlock.go:102]: Is this used

File: process/peer/process.go [hunk 10]
```diff
@@ -775,6 +775,7 @@ func (vs *validatorStatistics) updateShardDataPeerState(
 			return shardInfoErr
 		}
 
+		log.Debug("validator statistics ", "shHdr nonce ", h.Nonce, "genesisNonce", vs.genesisNonce)
```
Comment (SebastianMarian) [inline process/peer/process.go:778]: log.Trace?

File: epochStart/metachain/economics.go [hunk 11]
```diff
@@ -90,7 +90,7 @@ func (e *economics) ComputeEndOfEpochEconomics(
 	if metaBlock.DevFeesInEpoch == nil {
 		return nil, epochStart.ErrNilTotalDevFeesInEpoch
 	}
-	if !metaBlock.IsStartOfEpochBlock() || metaBlock.Epoch < 1 {
+	if !metaBlock.IsStartOfEpochBlock() || metaBlock.Epoch < e.genesisEpoch {
```
Comment (SebastianMarian) [inline epochStart/metachain/economics.go:93]: e.genesisEpoch + 1  ?

File: genesis/process/metaGenesisBlockCreator.go [hunk 12]
```diff
@@ -148,6 +148,11 @@ func createMetaGenesisAfterHardFork(
 	}
 	saveGenesisBodyToStorage(processors.txCoordinator, bodyHandler)
 
+	err = saveGenesisMetaToStorage(arg.Store, arg.Marshalizer, metaHdr)
```
Comment (SebastianMarian) [inline genesis/process/metaGenesisBlockCreator.go:151]: Should be critical? saveGenesisBodyToStorage method above just prints some warns inside and returns no error

File: update/process/metaBlock_test.go [hunk 13]
```diff
@@ -122,6 +122,8 @@ func TestMetaBlockCreator_CreateNewBlock(t *testing.T) {
 		SoftwareVersion:        []byte(""),
 		AccumulatedFees:        big.NewInt(0),
 		AccumulatedFeesInEpoch: big.NewInt(0),
+		DeveloperFees:          big.NewInt(0),
+		DevFeesInEpoch:         big.NewInt(0),
```
Comment (SebastianMarian) [inline update/process/metaBlock_test.go:126]: Is this used?

File: update/process/shardBlock_test.go [hunk 14]
```diff
@@ -163,6 +163,7 @@ func TestCreateNewBlock(t *testing.T) {
 		ReceiptsHash: []byte("receiptHash"), TxCount: 2,
 		MetaBlockHashes: [][]byte{metaHash}, MiniBlockHeaders: expectedMbHeaders,
 		AccumulatedFees: big.NewInt(0),
+		DeveloperFees:   big.NewInt(0),
```
Comment (SebastianMarian) [inline update/process/shardBlock_test.go:166]: DevFeesInEpoch is not needed to be set on 0 ?

File: genesis/process/metaGenesisBlockCreator.go (hunk 12 above)
Comment (sasurobert) [reply genesis/process/metaGenesisBlockCreator.go:151]: Yes. It is critical. As it is used in compute economics in the first epoch

File: update/process/metaBlock_test.go (hunk 13 above)
Comment (sasurobert) [reply update/process/metaBlock_test.go:126]: Yes. It is used

File: update/process/shardBlock_test.go (hunk 14 above)
Comment (sasurobert) [reply update/process/shardBlock_test.go:166]: It is only a test

File: update/process/metaBlock.go (hunk 9 above)
Comment (sasurobert) [reply update/process/metaBlock.go:102]: Yes. It is used

File: update/process/metaBlock.go (hunk 8 above)
Comment (sasurobert) [reply update/process/metaBlock.go:101]: This was not part of this PR

File: process/mock/blockTrackerMock.go [hunk 15]
```diff
@@ -381,6 +382,10 @@ func (btm *BlockTrackerMock) GetSelfNotarizedHeader(shardID uint32, offset uint6
 	btm.mutSelfNotarizedHeaders.RLock()
 	defer btm.mutSelfNotarizedHeaders.RUnlock()
 
+	if len(btm.selfNotarizedHeaders[core.MetachainShardId]) == 0 {
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply process/mock/blockTrackerMock.go:385]: It is only a mock. Used for tests

File: process/sync/storageBootstrap/baseStorageBootstrapper.go [hunk 16]
```diff
@@ -203,6 +205,13 @@ func (st *storageBootstrapper) applyHeaderInfo(hdrInfo bootstrapStorage.Bootstra
 		return err
 	}
 
+	if string(headerFromStorage.GetChainID()) != st.chainID {
+		log.Debug("cannot applyHeadInfo chainID missmatch",
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply process/sync/storageBootstrap/baseStorageBootstrapper.go:209]: done

File: process/peer/process.go (hunk 10 above)
Comment (sasurobert) [reply process/peer/process.go:778]: log is not needed anymore. deleted

File: epochStart/metachain/economics.go (hunk 11 above)
Comment (sasurobert) [reply epochStart/metachain/economics.go:93]: done

File: scripts/testnet/include/config.sh [hunk 17]
```diff
@@ -205,3 +205,19 @@ updateJSONValue() {
...
   sed -i "s,\"$key\": .*\$,\"$key\": $escaped_value\,," $filename
 }
+
+changeConfigForHardfork(){
+  pushd $TESTNETDIR/node/config
+
+  export FIRST_PUBKEY=$(cat nodesSetup.json | grep pubkey -m 1 | cut -c18-209)
+  updateTOMLValue config_observer.toml "PublicKeyToListenFrom" "\"$FIRST_PUBKEY\""
+  updateTOMLValue config_validator.toml "PublicKeyToListenFrom" "\"$FIRST_PUBKEY\""
+}
```
Comment (camilbancioiu) [inline scripts/testnet/include/config.sh:215]: Must `popd` if you `pushd`.

File: scripts/testnet/include/config.sh [hunk 18]
```diff
@@ -205,3 +205,19 @@ updateJSONValue() {
...
+
+copyBackConfigs(){
+  pushd $TESTNETDIR
+
+  echo "trying to copy-back the configs"
+  cp ./node/config/*.* $NODEDIR/config
+  cp $NODEDIR/config/config_validator.toml $NODEDIR/config/config.toml
+}
```
Comment (camilbancioiu) [inline scripts/testnet/include/config.sh:223]: Don't forget to `popd`.

File: scripts/testnet/include/config.sh [hunk 19]
```diff
@@ -205,3 +205,19 @@ updateJSONValue() {
 
   sed -i "s,\"$key\": .*\$,\"$key\": $escaped_value\,," $filename
 }
+
+changeConfigForHardfork(){
+  pushd $TESTNETDIR/node/config
+
+  export FIRST_PUBKEY=$(cat nodesSetup.json | grep pubkey -m 1 | cut -c18-209)
```
Comment (camilbancioiu) [inline scripts/testnet/include/config.sh:212]:
```suggestion
  export FIRST_PUBKEY=$(cat nodesSetup.json | grep pubkey -m 1 | sed -E 's/^.*"([0-9a-f]+)".*$/\1/g')
```
`cut` is too fragile, depends on exact character counts. The above `sed` should extract the hexadecimal number from the `grep`ped line independently of tabs / spaces / quotes, commas etc.

File: scripts/testnet/include/config.sh [hunk 20]
```diff
@@ -205,3 +205,19 @@ updateJSONValue() {
...
+  updateTOMLValue config_observer.toml "PublicKeyToListenFrom" "\"$FIRST_PUBKEY\""
+  updateTOMLValue config_validator.toml "PublicKeyToListenFrom" "\"$FIRST_PUBKEY\""
+}
+
+copyBackConfigs(){
+  pushd $TESTNETDIR
+
+  echo "trying to copy-back the configs"
```
Comment (camilbancioiu) [inline scripts/testnet/include/config.sh:220]: `echo` not needed?

File: scripts/testnet/start.sh [hunk 21]
```diff
@@ -11,6 +11,12 @@ source "$ELRONDTESTNETSCRIPTSDIR/include/build.sh"
 source "$ELRONDTESTNETSCRIPTSDIR/include/nodes.sh"
 source "$ELRONDTESTNETSCRIPTSDIR/include/tools.sh"
 
+# Load local overrides, .gitignored
+LOCAL_OVERRIDES="$ELRONDTESTNETSCRIPTSDIR/local.sh"
+if [ -f "$LOCAL_OVERRIDES" ]; then
+  source "$ELRONDTESTNETSCRIPTSDIR/local.sh"
+fi
+
```
Comment (camilbancioiu) [inline scripts/testnet/start.sh:19]: `variables.sh` already loads `local.sh`, did you need to override funtions in the `include/*.sh` files?

File: scripts/testnet/include/nodes.sh [hunk 22]
```diff
@@ -120,19 +124,23 @@ iterateOverValidators() {
...
-    for VALIDATOR_IN_SHARD in `seq $SHARD_VALIDATORCOUNT`; do
-      $callback $SHARD $VALIDATOR_INDEX
-      let VALIDATOR_INDEX++
+  (( max_shard_id=$SHARDCOUNT - 1 ))
+  for SHARD in $(seq 0 1 $max_shard_id); do
+    for _ in $(seq $SHARD_VALIDATORCOUNT); do
+      if [ $VALIDATOR_INDEX -ne $SKIP_VALIDATOR_IDX ]; then
+        $callback $SHARD $VALIDATOR_INDEX
+      fi
+      (( VALIDATOR_INDEX++ ))
```
Comment (camilbancioiu) [inline scripts/testnet/include/nodes.sh:133]: Why the replacement of `let` with `(( ))`? Not a problem, just asking. Replacement of back-ticks with `$()` is a good idea though.

File: scripts/testnet/include/nodes.sh [hunk 23]
```diff
@@ -55,15 +55,19 @@ iterateOverObservers() {
   let "max_shard_id=$SHARDCOUNT - 1"
   for SHARD in `seq 0 1 $max_shard_id`; do
     for OBSERVER_IN_SHARD in `seq $SHARD_OBSERVERCOUNT`; do
-      $callback $SHARD $OBSERVER_INDEX
+      if [ $OBSERVER_INDEX -ne $SKIP_OBSERVER_IDX ]; then
+        $callback $SHARD $OBSERVER_INDEX
+      fi
       let OBSERVER_INDEX++
```
Comment (camilbancioiu) [inline scripts/testnet/include/nodes.sh:61]: If you want to replace `let` with `(( ))`, it needs to be replaced here as well. Back-ticks with `$()` too.

File: scripts/testnet/include/nodes.sh [hunk 24]
```diff
@@ -55,15 +55,19 @@ iterateOverObservers() {
...
   done
 
   # Iterate over Metachain Observers
   local SHARD="metachain"
   for META_OBSERVER in `seq $META_OBSERVERCOUNT`; do
-    $callback $SHARD $OBSERVER_INDEX
+    if [ $OBSERVER_INDEX -ne $SKIP_OBSERVER_IDX ]; then
+      $callback $SHARD $OBSERVER_INDEX
+    fi
     let OBSERVER_INDEX++
```
Comment (camilbancioiu) [inline scripts/testnet/include/nodes.sh:71]: ...and here.

File: scripts/testnet/hardfork.sh [hunk 25]
```diff
@@ -0,0 +1,44 @@
...
+
+export ELRONDTESTNETSCRIPTSDIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" >/dev/null 2>&1 && pwd )"
+source "$ELRONDTESTNETSCRIPTSDIR/variables.sh"
+source "$ELRONDTESTNETSCRIPTSDIR/include/config.sh"
+
+# Load local overrides, .gitignored
+LOCAL_OVERRIDES="$ELRONDTESTNETSCRIPTSDIR/local.sh"
+if [ -f "$LOCAL_OVERRIDES" ]; then
+  source "$ELRONDTESTNETSCRIPTSDIR/local.sh"
+fi
```
Comment (camilbancioiu) [inline scripts/testnet/hardfork.sh:11]: `local.sh` is already loaded by `variables.sh`, is it needed here?

File: scripts/testnet/hardfork.sh [hunk 26]
```diff
@@ -0,0 +1,44 @@
...
+
+epoch=$1
+cmd=(printf "$(curl -d '{"epoch":'"$epoch"'}' -H 'Content-Type: application/json' http://127.0.0.1:$VALIDATOR_RES_PORT/hardfork/trigger)")
+"${cmd[@]}"
+
+echo " done curl"
+
+# change the setting from config.toml: AfterHardFork to true
+sed -i '/AfterHardFork/,/AfterHardFork/ s/false/true/' "$TESTNETDIR/node/config/config_validator.toml"
+sed -i '/AfterHardFork/,/AfterHardFork/ s/false/true/' "$TESTNETDIR/node/config/config_observer.toml"
```
Comment (camilbancioiu) [inline scripts/testnet/hardfork.sh:33]: Consider using the function `updateTOMLValue` from `include/config.sh`.

File: scripts/testnet/include/config.sh (hunk 17 above)
Comment (camilbancioiu) [reply scripts/testnet/include/config.sh:217]: Done, please check.

File: scripts/testnet/include/config.sh (hunk 18 above)
Comment (camilbancioiu) [reply scripts/testnet/include/config.sh:227]: Done, please check.

File: scripts/testnet/include/config.sh (hunk 19 above)
Comment (camilbancioiu) [reply scripts/testnet/include/config.sh]: Done, please check.

File: scripts/testnet/include/nodes.sh (hunk 23 above)
Comment (camilbancioiu) [reply scripts/testnet/include/nodes.sh]: Done, please check.

File: scripts/testnet/include/nodes.sh (hunk 24 above)
Comment (camilbancioiu) [reply scripts/testnet/include/nodes.sh]: Done, please check.

File: scripts/testnet/hardfork.sh (hunk 26 above)
Comment (camilbancioiu) [reply scripts/testnet/hardfork.sh]: Done, please check.

File: scripts/testnet/include/nodes.sh (hunk 22 above)
Comment (iulianpascalau) [reply scripts/testnet/include/nodes.sh:133]: Goland bash linter suggestion

File: scripts/testnet/include/nodes.sh (hunk 23 above)
Comment (iulianpascalau) [reply scripts/testnet/include/nodes.sh]: Changed only in a few occurrences. Indeed, we should decide how to refactor all scripts

File: scripts/testnet/start.sh (hunk 21 above)
Comment (iulianpascalau) [reply scripts/testnet/start.sh:19]: removed

File: scripts/testnet/hardfork.sh (hunk 25 above)
Comment (iulianpascalau) [reply scripts/testnet/hardfork.sh:11]: removed

**Pull request merged**